teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -136,32 +149,59 @@ Consider including folks that also work outside the WGs or subproject.
 
 ### User Experience (optional)
 
-<!--
-Consideration about the user experience. Depending on the area of change,
-users may be task and pipeline editors, they may trigger task and pipeline
-runs or they may be responsible for monitoring the execution of runs,
-via CLI, dashboard or a monitoring system.
-
-Consider including folks that also work on CLI and dashboard.
--->
+Support for additional command or flags in `tkn` CLI should be considered.
 
 ### Performance (optional)
 
 No impact on performance.
 
 ## Design Details
 
-<!--
-This section should contain enough information that the specifics of your
-change are understandable.  This may include API specs (though not always
-required) or even code snippets.  If there's any ambiguity about HOW your
-proposal will be implemented, this is the place to discuss them.
-
-If it's helpful to include workflow diagrams or any other related images,
-add them under "/teps/images/". It's upto the TEP author to choose the name
-of the file, but general guidance is to include at least TEP number in the
-file name, for example, "/teps/images/NNNN-workflow.jpg".
--->
+In this proposal we need to consider following cases:
+
+1. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks, but `finally` section is empty.
+
+    In this case a graceful termination would ack the same as a pipeline run cancellation.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: False, Reason: Cancelled } 
+      ```
```

Comment:
In this case, how would the `pipelineRun` state transition look like? 

`Running` ->  `PipelineRunStopping` -> `Cancelled` ?

Or 

`Running` -> `Cancelled` ?

Pros/Cons of this approach:

+ Aligns well with the cancellation terminology.
- Feels like introducing one more status `PipelineRunStopped` for what `PipelineRunCancelled` is designed to do. 

Changing this behavior to ignore `PipelineRunStopped` in case of no final tasks might align with overall goal of graceful termination but might confuse users while exiting `pipelineRun` with the original state.

I think this behavior make sense and would keep it as is, unless there is any different feedback/opinion.

Reply:
@pritidesai The expected state transition is:

`Running` -> `PipelineRunStopping` -> `Cancelled`

similarly as in failure scenario:

`Running` -> `PipelineRunStopping` -> `Failed`

The state `PipelineRunStopping` is used to indicate that all non-final, unscheduled tasks should be skipped and then final tasks should run.

In my opinion, that works in the graceful termination case very well.

Reply:
> similarly as in failure scenario:
> 
> `Running` -> `PipelineRunStopping` -> `Failed`

thanks @rafalbigaj, yes I agree. The only odd behavior I see is transitioning into `PipelineRunStopping` with `finally` tasks. So far, this state was only visible in case of the pipelines without `finally` tasks.